### ISSUE TITLE
Setup wizard needs retries to complete

### ISSUE THREAD

**Issue by revant** [frappe/frappe_docker#751]
Looks like I got to excited too. I saw the setup wizard and thought It was running. I can't get past the initial setup either. I feel this is probably related to the docker setup. 
Client: Docker Engine - Community  Version:           20.10.13
Docker Compose version: v2.2.3

![Screenshot from 2022-03-25 10-32-20](https://user-images.githubusercontent.com/17303371/160029923-28ba86da-1141-48fa-9d5e-e809347db66d.png)

![Screenshot from 2022-03-25 10-32-42](https://user-images.githubusercontent.com/17303371/160029884-be50a97f-f64d-4099-bd28-3850f3f8f59c.png)

```
...
example-queue-default-1  | 23:46:23 default: frappe.utils.background_jobs.execute_job(event=None, is_async=True, job_name='frappe.core.doctype.user.user.create_contact', kwargs={'user': <User: user1@example.org>, 'ignore_mandatory': True}, method='frappe.core.doctype.user.user.create_contact', site='app.example.org', user='Administrator') (06e10e0b-28d1-4709-a763-a99ec95f4b77)
example-queue-default-1  | 23:46:24 Traceback (most recent call last):
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/worker.py", line 1013, in perform_job
example-queue-default-1  |     rv = job.perform()
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/job.py", line 709, in perform
example-queue-default-1  |     self._result = self._execute()
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/job.py", line 732, in _execute
example-queue-default-1  |     result = self.func(*self.args, **self.kwargs)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/background_jobs.py", line 115, in execute_job
example-queue-default-1  |     method(**kwargs)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 997, in create_contact
example-queue-default-1  |     contact.insert(ignore_permissions=True, ignore_links=ignore_links, ignore_mandatory=ignore_mandatory)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 234, in insert
example-queue-default-1  |     self._validate_links()
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 837, in _validate_links
example-queue-default-1  |     frappe.throw(_("Could not find {0}").format(msg),
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 444, in throw
example-queue-default-1  |     msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 423, in msgprint
example-queue-default-1  |     _raise_exception()
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 378, in _raise_exception
example-queue-default-1  |     raise raise_exception(msg)
example-queue-default-1  | frappe.exceptions.LinkValidationError: Could not find User Id: user1@example.org
example-queue-default-1  | 
example-queue-default-1  | 23:46:24 default: frappe.utils.background_jobs.execute_job(event=None, is_async=True, job_name='frappe.core.doctype.user.user.update_gravatar', kwargs={'name': 'user1@example.org'}, method='frappe.core.doctype.user.user.update_gravatar', site='app.example.org', user='Administrator') (c9fe1f6c-8e8a-425b-ac09-c4a7dbd210eb)
example-queue-default-1  | 23:46:24 default: Job OK (c9fe1f6c-8e8a-425b-ac09-c4a7dbd210eb)
example-queue-default-1  | 23:46:24 Result is kept for 500 seconds
example-queue-default-1  | 23:46:24 default: frappe.utils.background_jobs.execute_job(event=None, is_async=True, job_name='frappe.core.doctype.user.user.create_contact', kwargs={'user': <User: user1@example.org>, 'ignore_mandatory': True}, method='frappe.core.doctype.user.user.create_contact', site='app.example.org', user='Administrator') (3921fd29-11c1-4d9e-a929-521bdb5032ca)
example-queue-default-1  | 23:46:25 Traceback (most recent call last):
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/worker.py", line 1013, in perform_job
example-queue-default-1  |     rv = job.perform()
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/job.py", line 709, in perform
example-queue-default-1  |     self._result = self._execute()
example-queue-default-1  |   File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/rq/job.py", line 732, in _execute
example-queue-default-1  |     result = self.func(*self.args, **self.kwargs)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/background_jobs.py", line 115, in execute_job
example-queue-default-1  |     method(**kwargs)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/core/doctype/user/user.py", line 997, in create_contact
example-queue-default-1  |     contact.insert(ignore_permissions=True, ignore_links=ignore_links, ignore_mandatory=ignore_mandatory)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 234, in insert
example-queue-default-1  |     self._validate_links()
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/model/document.py", line 837, in _validate_links
example-queue-default-1  |     frappe.throw(_("Could not find {0}").format(msg),
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 444, in throw
example-queue-default-1  |     msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 423, in msgprint
example-queue-default-1  |     _raise_exception()
example-queue-default-1  |   File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 378, in _raise_exception
example-queue-default-1  |     raise raise_exception(msg)
example-queue-default-1  | frappe.exceptions.LinkValidationError: Could not find User Id: user1@example.org
example-queue-default-1  | 
example-queue-default-1  | 23:46:25 default: frappe.utils.background_jobs.execute_job(event=None, is_async=True, job_name='frappe.core.doctype.user.user.update_gravatar', kwargs={'name': 'user1@example.org'}, method='frappe.core.doctype.user.user.update_gravatar', site='app.example.org', user='Administrator') (c4d488fc-87ba-4020-afd2-cd9fd9c07248)
example-queue-default-1  | 23:46:25 default: Job OK (c4d488fc-87ba-4020-afd2-cd9fd9c07248)
example-queue-default-1  | 23:46:25 Result is kept for 500 seconds
example-backend-1        | [2022-03-24 23:46:51 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:8)
example-backend-1        | [2022-03-24 23:46:51 +0000] [8] [ERROR] Error handling request /api/method/frappe.desk.page.setup_wizard.setup_wizard.setup_complete
...
```

_Originally posted by @1ubuntuuser in https://github.com/frappe/frappe_docker/issues/715#issuecomment-1078522655_

**Comment by revant** [comment]
I think the `frappe.exceptions.LinkValidationError: Could not find User Id: user1@example.org` is misleading. It has nothing to do with wizard refresh not happening.

I made following change to the main `compose.yaml` and it somehow worked.

```yaml
  backend:
    <<: *backend_defaults
    volumes:
      - sites:/home/frappe/frappe-bench/sites
      - assets:/home/frappe/frappe-bench/sites/assets:ro
    command:
      - /home/frappe/frappe-bench/env/bin/gunicorn
      - --bind=0.0.0.0:8000
      - --worker-tmp-dir=/dev/shm
      - --threads=4
      - --workers=2
      - --worker-class=gthread
      - --log-file=-
      - --timeout=120
      - --preload
      - frappe.app:application
```

I tried the default command that was used for gunicorn in previous containers by overriding it in current setup. It had workers, threads, timeout, preload. I think it has something to do with containers+python+gunicorn. We referred this blog https://pythonspeed.com/articles/gunicorn-in-docker when initial command was used. 

@vrslev @xelfe @1ubuntuuser
Can anyone confirm this.

**Comment by revant** [comment]
Related issue due to old command not used.

https://github.com/frappe/frappe_docker/issues/747

**Comment by vrslev** [comment]
@revant Should we make this the default? Also, how many workers and threads should we use?

**Comment by revant** [comment]
Can you confirm this? just confirm with default values. I tried it twice I guess it is the solution.

The blog article is informative go through it. In brief,

- gunicorn should use /dev/shm instead of /tmp
- more than one worker and thread needed to handle heartbeat. We can keep it to 2 workers and 4 threads just to follow the article?
- log to stdout
- worker-class, as the article mentions gthread was used.
- additional configuration in old command
  - preload https://docs.gunicorn.org/en/stable/settings.html#preload-app
  - timeout https://docs.gunicorn.org/en/stable/settings.html#timeout default is 30, old command made it 120
 
I tried the overriden command again with just `--timeout 120`. I think for this case of wizard extending the timeout to 120 from default 30 is the actual solution.

the native setup with bench also defaults to `120`.

https://github.com/frappe/bench/blob/2bd3cfae075aa614d3ccf2832cc5de401fc8dc3e/bench/config/supervisor.py#L34 

@vrslev should we follow the article for the config? Or just set the timeout to 120?

**Comment by vrslev** [comment]
I like this setup. No logs to stdout (we have nginx, why do useless work?), all things from the article.

```bash
gunicorn -b 0.0.0.0:8000 \ 
   --threads 4 \
   --workers 2 \
   --worker-class gthread \
   --worker-tmp-dir /dev/shm \ 
   --timeout 120 \
   --preload 
   frappe.app:application
```

**Comment by vrslev** [comment]
> Can you confirm this? just confirm with default values. I tried it twice I guess it is the solution.

Yes! Your solution works for me.

**Comment by revant** [comment]
> I like this setup. No logs to stdout (we have nginx, why do useless work?), all things from the article.
> 
> ```shell
> gunicorn \
>    --bind 0.0.0.0:8000 \ 
>    --threads 4 \
>    --workers 2 \
>    --worker-class gthread \
>    --worker-tmp-dir /dev/shm \ 
>    --timeout 120 \
>    --preload 
>    frappe.app:application
> ```

We'll make this default. Back to what it was. Anyway anything else can be overridden.

**Comment by vrslev** [comment]
Can you make the change?

**Comment by 1ubuntuuser** [comment]
I can confirm this is fixed when tested with v13.26.0.

Great work guys!

**Comment by xelfe** [comment]
It's working like a charm !
